src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -835,9 +835,22 @@ private static SelectedPropertiesNode ProcessSubExpand(string nodeName, Selected
         /// <returns>The generated SelectedPropertiesNode.</returns>
         private static SelectedPropertiesNode CombineSelectAndExpandResult(IEnumerable<string> selectList, IEnumerable<SelectedPropertiesNode> expandList)
         {
-            List<string> rawSelect = selectList.ToList();
-            rawSelect.RemoveAll(expandList.Select(m => m.nodeName).Contains);
+            HashSet<string> expandSet = new HashSet<string>();
+            foreach(SelectedPropertiesNode propNode in expandList)
+            {
+                expandSet.Add(propNode.nodeName);
+            }
 
+            List<string> rawSelect = new List<string>();
+
+            foreach(string name in selectList)
+            {
+                if (!expandSet.Contains(name))
+                {
+                    rawSelect.Add(name);
+                }
+            }
+                       
             if (rawSelect.Count == 0 && expandList.All(n => n.IsEntireSubtree()))

```

Comment:
>expandList.All(n => n.IsEntireSubtree() [](start = 40, length = 39)

rather than re-iterate through expandlist, just add a boolean property "isEntireSubtree" that is initialized to true and changed to false as you iterate through the expandList on 839 building expandSet.